### DISCUSSION TITLE
use actix web from other dependencies

### DISCUSSION THREAD

**Discussion by shishamo** [actix/actix-web#3105] [Q&A] [answered]
I would like to use actix web from a common dependency but i have an error as below
`could not find "actix_web" in the list of imported crates`
`actix_web_codegen`
`proc_macro main`

Here is my project tree
```
|-- common
| |-- src
| | |-- external_libs
| | | |-- mod.rs
| | |-- lib.rs
| |
| |-- cargo.toml
|
|
|-- project
| |-- app
| | |-- src
| | | |-- lib.rs
| | |-- cargo.toml
| |
| |-- api
| | |--cargo.toml
```

the `mod.rs` file in common external_libs dir
```rs
pub use actix_files;
```

the `lib.rs` file in common dir
```rs
mod external_libs;

pub use external_libs::*;
```

the cargo.toml file in common dir
```rs
[package]
name = "common"

[dependencies]
actix-web = "^4.3"
```

the cargo file in project dir
```rs
members = ["app/"]
```

the cargo file in project dir
```rs
[package]
name = "app"

[dependencies]
common = { path = "../../common/" }
```

and the lib.rs in app dir
```rs
use common::actix_web;

#[actix_web::main]
async fn main() -> Result<()> {
   ....
}
```

Rust analyzer can jump inside Actix web so looks like the dependency can be find but the compiler fails....

Could someone help me please? T_T

**Comment by shishamo** [comment]
any idea? 😂

**Comment by robjtede** [comment, marked as answer]
Macros can't resolve dependencies like this. You'll need to:
- avoid the attribute macro for this case
- use `#[::common::actix_web::rt::main(system = "::common::actix_web::rt::System")]`

**Comment by shishamo** [comment]
thanks a lot @robjtede ^^

btw how can i use the `#[actix_web::test]` in that case?

**Reply by shishamo** [reply]
it looks like it is working in using as below
`#[::static_common::actix_web::rt::test(system = "::static_common::actix_web::rt::System")]`

**Comment by shishamo** [comment]
but i have another issue this the get attribute macro
`#[get("/endpoint")]`

@robjtede 
how can i use it in that way?

**Reply by robjtede** [reply]
The routing macros currently have no support for alternative crate paths. You'll have to use manual routing.

**Reply by shishamo** [reply]
i see ~
thanks a lot!
